### PR TITLE
Add Alternate,Companion, Outlier PGM Testing and Evalution

### PR DESCRIPTION
**Pull request by sara-02**
First iteration.
- [x] Depends on #45 to be merged first

### CODE REVIEW

File: analytics_platform/kronos/uranus/src/alternate_testing.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
+# NOTE: Works only with S3DataStore
+from __future__ import division
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/alternate_testing.py]: This does not work with Python3, need to remove all __future__ imports.

File: analytics_platform/kronos/uranus/src/companion_testing.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
+# NOTE: Works only with S3DataStore
+from __future__ import division
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/companion_testing.py]: Same here.

File: analytics_platform/kronos/uranus/src/super_class.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    def load_package_list(self, training_url):
+        """Generate the aggregated manifest list for a given ecosystem.
+
+        :param input_manifest_data_store: The Data store to pick the manifest files from.
+        :param additional_path: The directory to pick the manifest files from."""
+
+        input_manifest_data_store, additional_path = self.get_input_data_store(
+            training_url)
+        manifest_filenames = input_manifest_data_store.list_files(
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/super_class.py]: Use `os.path.join` here.

File: analytics_platform/kronos/uranus/src/super_class.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    AWS_S3_ACCESS_KEY_ID,
+    AWS_S3_SECRET_ACCESS_KEY,
+    KRONOS_MODEL_PATH,
+    KRONOS_SCORING_REGION)
+
+import os
+import time
+import sys
+sys.path.insert(0, SPARK_HOME_PATH + "/python")
+sys.path.insert(0, os.path.join(SPARK_HOME_PATH, "python/lib/" + PY4J_VERSION))
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/super_class.py]: This can be taken care of by setting the PYTHONPATH in the Dockerfile.

Comment (sara-02) [reply analytics_platform/kronos/uranus/src/super_class.py]: But will still be needed when running locally.

Comment (pkajaba) [reply analytics_platform/kronos/uranus/src/super_class.py]: so can't you use virtual env then? This is rather a messy way how to deal with `PYTHONPATH` and I agree with @rootAvish here.

**Comment by miteshvp** [comment]
@sara-02 can we run a shell using `subprocess` rather than current approach. WDYT?

**Comment by miteshvp** [comment]
Otherwise LGTM

File: analytics_platform/kronos/uranus/src/alternate_testing.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        :param additional_path: The directory where test data is loaded from."""
+
+        input_filename = additional_path + \
+            URANUS_OUTPUT_PATH + "reverse_dict.json"
+        self.unique_package_dict = self.load(
+            input_data_store,
+            input_filename)
+        self.test_set_len = len(self.unique_package_dict)
+
+        input_filename = additional_path + \
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/alternate_testing.py]: use `os.path.join` here.

File: analytics_platform/kronos/uranus/src/alternate_testing.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            URANUS_OUTPUT_PATH + "freq_4.json"
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/alternate_testing.py]: here as well.

File: analytics_platform/kronos/uranus/src/companion_outlier_testing.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        self.unique_items_len = 0
+
+    def load_attributes(self, input_data_store, additional_path):
+        """Load the attributes of the class object.
+
+        :param input_data_store: The place to fetch the data from.
+        :param additional_path: The directory to pick the manifest files from."""
+
+        pickle_filename = "comp_test_set.pickle"
+        complete_output_filename = additional_path + \
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/companion_outlier_testing.py]: Here as well.

File: analytics_platform/kronos/uranus/src/companion_outlier_testing.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            URANUS_OUTPUT_PATH + pickle_filename
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/companion_outlier_testing.py]: Same.

File: analytics_platform/kronos/uranus/src/generate_test_data.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        self.comp_test_set = set()
+
+    def load_package_list(self, input_data_store, additional_path):
+        """Generate the aggregated manifest list for a given ecosystem.
+
+        :param input_data_store: The Data store to pick the manifest files from.
+        :param additional_path: The directory to pick the manifest files from."""
+
+        manifest_filenames = input_data_store.list_files(
+            additional_path + URANUS_INPUT_RAW_PATH)
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/generate_test_data.py]: Same here.

File: analytics_platform/kronos/uranus/src/generate_test_data.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+    def save_json_file(output_data_store, additional_path, filename, contents):
+        """Save the contents as a JSON file at the given datastore.
+
+        :param output_data_store: The datastore where json will be saved.
+        :param additional_path: The directory where the json will be saved.
+        :param filename: The name under which json data will be saved.
+        :param contents: The json data to be saved."""
+
+        output_filename = additional_path + \
+            URANUS_OUTPUT_PATH + filename
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/generate_test_data.py]: same.

File: analytics_platform/kronos/uranus/src/generate_test_data.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+    @staticmethod
+    def save_pickle_file(output_data_store, additional_path, pickle_filename):
+        """Save the pickle file at the given datastore.
+
+        :param output_data_store: The datastore where json will be saved.
+        :param additional_path: The directory where the json will be saved.
+        :param pickle_filename: The locally stored pickle file to be saved."""
+
+        complete_filename = additional_path + \
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/generate_test_data.py]: Here as well.

File: analytics_platform/kronos/uranus/src/generate_test_data.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+            pickle_filename=pickle_filename)
+
+    @staticmethod
+    def pickle_dump(pickle_filename, contents):
+        """Save the contents as a pickle file.
+
+        :param  contents: The data to pickle
+        :param pickle_filename: The file name under which data is pickled."""
+
+        with open('/tmp/' + pickle_filename, 'wb') as handle:
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/generate_test_data.py]: Here as well.

File: analytics_platform/kronos/uranus/src/super_class.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        return False
+
+    def load_search_set(self, input_data_store, additional_path):
+        """Load the manifets search set from the given source.
+
+        :param input_data_store: The datastore where search set is present.
+        :param additional_path: The directory where search set is loaded from."""
+
+        pickle_filename = "search_set.pickle"
+        input_filename = additional_path + \
```
Comment (rootAvish) [inline analytics_platform/kronos/uranus/src/super_class.py]: Here as well.

File: util/data_store/s3_data_store.py [hunk 14]
```diff
@@ -73,6 +74,19 @@ def write_json_file(self, filename, contents):
         self.s3_resource.Object(self.bucket_name, filename).put(Body=json.dumps(contents))
         return None
 
+    def write_pickle_file(self, complete_filename, pickle_filename):
+        """Write Pickle file into S3 bucket"""
+
+        self.s3_resource.Object(self.bucket_name, complete_filename).put(
+            Body=open('/tmp/' + pickle_filename, 'rb'))
```
Comment (rootAvish) [inline util/data_store/s3_data_store.py]: Here as well.

File: evaluation_platform/uranus/src/super_class.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        return False
+
+    def load_search_set(self, input_data_store, additional_path):
+        """Load the manifets search set from the given source.
+
+        :param input_data_store: The datastore where search set is present.
+        :param additional_path: The directory where search set is loaded from."""
+
+        pickle_filename = "search_set.pickle"
+        input_filename = additional_path + \
```
Comment (rootAvish) [inline evaluation_platform/uranus/src/super_class.py]: use `os.path.join` here.

File: analytics_platform/kronos/deployment/rest_api.py [hunk 16]
```diff
@@ -128,6 +130,23 @@ def predict_and_score():
     return flask.jsonify(response)
 
 
+@app.route('/api/v1/schemas/kronos_evaluation', methods=['POST'])
+def submit_kronos_evaluation():
+    app.logger.info("Submitting the evaluation job")
+    response = {"message": "Failed to load model, Kronos Region not available"}
```
Comment (rootAvish) [inline analytics_platform/kronos/deployment/rest_api.py]: This message will not get overwritten, which is something I think you wanted.

**Pull request merged**